streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -447,6 +459,49 @@ private void removePartitions() {
         tasks.clear();
     }
 
+    public PartitionGrouper partitionGrouper() {
+        return partitionGrouper;
+    }
+
+    private void ensureCopartitioning(Collection<Set<String>> copartitionGroups) {
+        for (Set<String> copartitionGroup : copartitionGroups) {
+            ensureCopartitioning(copartitionGroup);
+        }
+    }
+
+    private void ensureCopartitioning(Set<String> copartitionGroup) {
+        int numPartitions = -1;
+
+        for (String topic : copartitionGroup) {
```

Comment:
Thinking about it twice, maybe this requirement is a bit too restrict: for example we may have a low-traffic topic with N partitions and a high-traffic topic with N * M partitions and we still want to join them as partition1 with partition1-M, etc. Admittedly this will bring other problems when we change number of partitions, but maybe worth considering.